indra/newview/llagent.cpp

```diff
@@ -1453,48 +1453,69 @@ LLVector3 LLAgent::getReferenceUpVector()
 	return up_vector;
 }
 
-
 // Radians, positive is forward into ground
 //-----------------------------------------------------------------------------
 // pitch()
 //-----------------------------------------------------------------------------
 void LLAgent::pitch(F32 angle)
 {
-	// don't let user pitch if pointed almost all the way down or up
+    if (fabs(angle) <= 1e-4)
+        return;
 
-	// A dot B = mag(A) * mag(B) * cos(angle between A and B)
-	// so... cos(angle between A and B) = A dot B / mag(A) / mag(B)
-	//                                  = A dot B for unit vectors
+    LLCoordFrame newCoordFrame(mFrameAgent);
+    newCoordFrame.pitch(angle);
 
-	LLVector3 skyward = getReferenceUpVector();
+    // don't let user pitch if rotated 180 degree around the vertical axis
+    if ((newCoordFrame.getXAxis()[VX] * mFrameAgent.getXAxis()[VX] < 0) &&
+        (newCoordFrame.getXAxis()[VY] * mFrameAgent.getXAxis()[VY] < 0))
+        return;
 
-	// clamp pitch to limits
-	if (angle >= 0.f)
-	{
-		const F32 look_down_limit = 179.f * DEG_TO_RAD;
-		F32 angle_from_skyward = acos(mFrameAgent.getAtAxis() * skyward);
-		if (angle_from_skyward + angle > look_down_limit)
-		{
-			angle = look_down_limit - angle_from_skyward;
-		}
-	}
-	else if (angle < 0.f)
-	{
-		const F32 look_up_limit = 5.f * DEG_TO_RAD;
-		const LLVector3& viewer_camera_pos = LLViewerCamera::getInstance()->getOrigin();
-		LLVector3 agent_focus_pos = getPosAgentFromGlobal(gAgentCamera.calcFocusPositionTargetGlobal());
-		LLVector3 look_dir = agent_focus_pos - viewer_camera_pos;
-		F32 angle_from_skyward = angle_between(look_dir, skyward);
-		if (angle_from_skyward + angle < look_up_limit)
-		{
-			angle = look_up_limit - angle_from_skyward;
-		}
-	}
+    // don't let user pitch if pointed almost all the way down or up
+    if (gAgentCamera.getCameraMode() == CAMERA_MODE_THIRD_PERSON ||
```

Comment:
Why did yu add this cheack? Won't this negatively affect avatar customization?

Reply:
Good point.

Reply:
Because each preset requires different checks

Reply:
The avatar customization mode won't be affected because it doesn't use vertical angles close to 1 and 179

Reply:
And CAMERA_MODE_FOLLOW? That's a mode where script overrides control over camera and otherwise acts as a CAMERA_MODE_THIRD_PERSON, doesn't appear to affect mouselook

Reply:
I checked and it behaves exactly like CAMERA_MODE_THIRD_PERSON

Reply:
With CAMERA_MODE_THIRD_PERSON your camera could rotate any directions without any action from your side
So this bug doesn't make sense for CAMERA_MODE_THIRD_PERSON

When the vertical angle takes critical values 1 or 179, camera could rotate, and nobody treats this like a bug